File: cpp_solutions/826.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        int difficultyIdx = -1;
        int workerIdx = 0;
        int n = difficulty.size();
        
        vector<pair<int, int>> diffAndProfit;
        for (int i = 0; i < difficulty.size(); ++i) {
            diffAndProfit.push_back({difficulty[i], profit[i]});
        }
        sort(diffAndProfit.begin(), diffAndProfit.end());
        sort(worker.begin(), worker.end());
        
        int maxProfit = 0;
        int res = 0;
        for (auto& workerCap : worker) {
            while (difficultyIdx + 1 < n && diffAndProfit[difficultyIdx + 1].first <= workerCap) {
                difficultyIdx++;
                maxProfit = max(maxProfit, diffAndProfit[difficultyIdx].second);
            }
            res += maxProfit;
        }
        return res;
    }
};

File: rust_solutions/826.rs
impl Solution {
    pub fn max_profit_assignment(difficulty: Vec<i32>, profit: Vec<i32>, mut worker: Vec<i32>) -> i32 {
        let n = difficulty.len();
        let mut v = vec![];

        for i in 0..n {
            v.push((difficulty[i], profit[i]));
        }

        v.sort();

        let mut res = 0;
        let mut max_profit = 0;
        let mut idx = 0;
        worker.sort();
        let m = worker.len();
        for i in 0..m {
            while idx < n && worker[i] >= v[idx].0 {
                max_profit = i32::max(max_profit, v[idx].1);
                idx += 1;
            }
            res += max_profit;
        }
        return res;
    }
}
